src/permissions.rs

```diff
@@ -7,178 +7,315 @@ use ansi_term::Style;
     return Ok(());
   }
 
   pub fn revoke_read(&self) -> DenoResult<()> {
-    self.allow_read.store(false, Ordering::SeqCst);
+    self.allow_read.revoke();
     return Ok(());
   }
 
   pub fn revoke_write(&self) -> DenoResult<()> {
-    self.allow_write.store(false, Ordering::SeqCst);
+    self.allow_write.revoke();
     return Ok(());
   }
 
   pub fn revoke_net(&self) -> DenoResult<()> {
-    self.allow_net.store(false, Ordering::SeqCst);
+    self.allow_net.revoke();
     return Ok(());
   }
 
   pub fn revoke_env(&self) -> DenoResult<()> {
-    self.allow_env.store(false, Ordering::SeqCst);
+    self.allow_env.revoke();
     return Ok(());
   }
+}
 
-  pub fn default() -> Self {
-    Self {
-      allow_read: AtomicBool::new(false),
-      allow_write: AtomicBool::new(false),
-      allow_env: AtomicBool::new(false),
-      allow_net: AtomicBool::new(false),
-      allow_run: AtomicBool::new(false),
-      ..Default::default()
+/// Quad-state value for representing user input on permission prompt
+#[derive(Debug, Clone)]
+pub enum PromptResult {
+  AllowAlways = 0,
+  AllowOnce = 1,
+  DenyOnce = 2,
+  DenyAlways = 3,
+}
+
+impl PromptResult {
+  /// If value is any form of deny this will error with permission_denied
+  pub fn check(&self) -> DenoResult<()> {
+    match self {
+      PromptResult::DenyOnce => Err(permission_denied()),
+      PromptResult::DenyAlways => Err(permission_denied()),
+      _ => Ok(()),
     }
   }
 }
 
-fn permission_prompt(message: &str) -> DenoResult<()> {
-  if !atty::is(atty::Stream::Stdin) || !atty::is(atty::Stream::Stderr) {
-    return Err(permission_denied());
-  };
-  let msg = format!("⚠️  Deno requests {}. Grant? [yN] ", message);
+fn permission_prompt(message: &str) -> DenoResult<PromptResult> {
+  let msg = format!("⚠️  Deno requests {}. Grant? [a/y/n/d (a = allow always, y = allow once, n = deny once, d = deny always)] ", message);
   // print to stderr so that if deno is > to a file this is still displayed.
   eprint!("{}", Style::new().bold().paint(msg));
-  let mut input = String::new();
-  let stdin = io::stdin();
-  let _nread = stdin.read_line(&mut input)?;
-  let ch = input.chars().next().unwrap();
-  let is_yes = ch == 'y' || ch == 'Y';
-  if is_yes {
-    Ok(())
-  } else {
-    Err(permission_denied())
+  loop {
+    let mut input = String::new();
+    let stdin = io::stdin();
+    let _nread = stdin.read_line(&mut input)?;
+    let ch = input.chars().next().unwrap();
+    match ch.to_ascii_lowercase() {
+      'a' => return Ok(PromptResult::AllowAlways),
+      'y' => return Ok(PromptResult::AllowOnce),
+      'n' => return Ok(PromptResult::DenyOnce),
+      'd' => return Ok(PromptResult::DenyAlways),
+      _ => {
+        // If we don't get a recognized option try again.
+        let msg_again = format!("Unrecognized option '{}' [a/y/n/d (a = allow always, y = allow once, n = deny once, d = deny always)] ", ch);
+        eprint!("{}", Style::new().bold().paint(msg_again));
+      }
```

Comment:
Cool - this looks great - but how to test it?

Reply:
Something like 'yes | deno ./test.ts ' ? And the test ask for permission twice. No?

Reply:
we can test it externally calling it by this kind of command: `y | deno ./permission_test.ts` and in the test you ask for permission twice for example. But i think it's more a workaround.

Reply:
I think there is a permission_prompt.py file which could be amended...

Reply:
My solution at this point is to have two calls to the same operation, and one of them wrapped in a try-catch with a distinct console output `First check failed`.
```typescript
try {
  readFileSync("package.json");
} catch (e) {
  console.log("First check failed");
}
readFileSync("package.json");
```
That way I can test combinations of allow once/deny once, and allow always/deny always.